doc: Bluetooth: Mesh: documentation for security toolbox and key importer (#20928)
nrfconnect/sdk-nrf: doc/nrf/protocols/bt/bt_mesh/configuring.rst

  * Ambient light level gain

Adding up all entries, it is worth setting the cache size to minimum 71.

Security toolbox
----------------

> Zephyr's Mesh security toolbox implementation does not include encryption and authentication functionality, such as CMAC, AES-CCM, and HMAC-SHA-256.
> The third-party crypto solutions are used instead.

Reviewer: Just a bit more concise and avoids negation.
```suggestion
Zephyr's Mesh security toolbox implementation uses third-party crypto libraries APIs (such as CMAC, AES-CCM, and HMAC-SHA-256) for implementing encryption and authentication functionality.
```

* The following options are available:

>   * :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA` - Enables use of the Mbed TLS PSA API based security toolbox (default option).

Reviewer: Good to include a link for MbedTSL doc landing page
```suggestion
  * :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA` - Enables use of the [Mbed TLS](url) PSA API based security toolbox (default option).
```

>   * :kconfig:option:`CONFIG_BT_MESH_USES_TFM_PSA` - Enables use of the TF-M PSA API based security toolbox (default option for platforms that support TF-M).

Reviewer: Same as above.

>   * :kconfig:option:`CONFIG_BT_MESH_USES_TINYCRYPT` - Enables use of Tinycrypt-based security toolbox.

Reviewer: Same as above.

Author: There is no link to Tinycrypt

>     Zephyr's Mesh operates with open value of keys, including storing them in the persistent memory.

Reviewer 2: ```suggestion
    Zephyr's Mesh operates with open key values, including storing them in the persistent memory.
```
Does this go wrong if we say open key values?

Author: That's ok to me

>     Zephyr's Mesh operates with open value of keys, including storing them in the persistent memory.

Reviewer: > open key values

What does it mean? I didn't understand ...

Author: open key value - 128 bits key
key ID - 16 bits ID from PSA API (it is associated with 128 bits key that is in secure domain)

>     The Tinycrypt-based solution has worse security materials protection compared to others and not recommended for future designs.

Reviewer 2: ```suggestion
    The Tinycrypt-based solution has worse security materials protection compared to others and is not recommended for future designs.
```

>     The Tinycrypt-based solution has worse security materials protection compared to others and not recommended for future designs.

Reviewer: Why it is worse? It will be nice to add one-liner explannation.

```suggestion
    The Tinycrypt-based solution has worse security materials protection compared to others  due to ... ... . Therefore, usage of Tinycrypt is not recommended for future designs.
```

>     The Tinycrypt-based solution has worse security materials protection compared to others due to keeping keys in memory in open form. Therefore, usage of Tinycrypt is not recommended for future designs.

Reviewer 2: ```suggestion
    The Tinycrypt-based solution has worse security materials protection compared to others due to keeping keys in memory in open form.
    Tinycrypt is not recommended for future designs.
```
Remember, all sentences on separate lines.

> The Bluetooth Mesh security toolbox based on the PSA API does not operate with open key values.

Reviewer: I think it is worth to use "key values in RAM' instead of saying "open key values"?

Author: but it keeps keys value in Flash and RRAM too. I would prefer to follow this formulation "open key value". It will help misinterpretation due to different memory type.

> After Bluetooth Mesh gets open key value, it immediately imports the keys into the crypto library and receives the unique key identifier.

Reviewer 2: ```suggestion
After Bluetooth Mesh receives an open key value, it immediately imports the key into the crypto library and receives the unique key identifier.
```
Or could it be ... sends it (the key value) to the crypto library...?
Or even ... it is immediately sent to the crypto library...

The key identifiers are used in the security toolbox and stored in the persistent memory.

> The crypto library is responsible for storing of the open key values in the Internal Trusted Storage (ITS).

Reviewer: ```suggestion
The crypto library is responsible for storing of the open key values in the Internal Trusted Storage ([ITS](url)).
```

> Bluetooth Mesh data structures based on Tinycrypt and based on the PSA API as well as images of these structures stored in the persistent memory are not compatible due to different key representations.

Reviewer: suggestion for readability.

```suggestion
Bluetooth Mesh data structures based on Tinycrypt and based on the PSA API, as well as images of these structures stored in the persistent memory, are not compatible due to different key representations.
```

> Bluetooth Mesh data structures based on Tinycrypt and based on the PSA API, as well as images of these structures stored in the persistent memory, are not compatible due to different key representations.

Reviewer 2: ```suggestion
Bluetooth Mesh data structures based on Tinycrypt and the PSA API, as well as images of these structures stored in the persistent memory, are not compatible due to different key representations.
```

is to be unprovisioned first and reprovisioned after update again.

> If the image is changed over Mesh DFU it is recommended to use :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`.

Author: Thanks for simplification. Really, this subclause caused some difficulties during writing. However, some sense has been lost a bit. I rephrased it a bit, basing on your suggestion.

> When a provisioned device updates its image using the Tinycrypt-based toolbox on image with the PSA API based toolbox, it needs to be unprovisioned first and reprovisioned after the update.
> If the image is changed over Mesh DFU, it is recommended to use :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`.

Reviewer: Shouldn't this be other way round?

```suggestion
When a provisioned device updates its firmware binary from the Tinycrypt-based toolbox to firmware binary that uses the PSA API based toolbox, a provisioned device must be unprovisioned first and reprovisioned after the update. If not <what happens... please state?> ...
If the image is changed over Mesh DFU, it is recommended to use :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`.
```

> When a provisioned device updates its image using the Tinycrypt-based toolbox on image with the PSA API based toolbox, it needs to be unprovisioned first and reprovisioned after the update.
> If the image is changed over Mesh DFU, it is recommended to use :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`.

Reviewer: ```suggestion
When a provisioned device updates its image using the Tinycrypt-based toolbox on image with the PSA API based toolbox, it needs to be unprovisioned first and reprovisioned after the update.
If the image is changed over Mesh DFU, it is recommended to use :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV` in [dfu_target](https://github.com/nrfconnect/sdk-nrf/blob/129d7fc6844ed24821f1d741a01131f1bbe58dd9/samples/bluetooth/mesh/dfu/common/src/dfu_target.c#L24) module.
```

Author: I guess documentation constructor shows definition correctly in the final documentation.

> If not then a device cannot restore data from the persistent memory after firmware update.

Reviewer 2: ```suggestion
The device cannot restore data from the persistent memory after firmware update.
```

> However, a provisioned device can update its image using the Tinycrypt-based toolbox on image with the PSA API based toolbox without being unprovisioned.
> The :kconfig:option:`CONFIG_BT_MESH_KEY_IMPORTER` Kconfig option enables the key importer functionality.

Reviewer: ```suggestion
However, a provisioned device can update its firmware image from the Tinycrypt-based toolbox to firmware image that uses the PSA API based toolbox without requiring to be  unprovisioned by using key importer functionality.
The :kconfig:option:`CONFIG_BT_MESH_KEY_IMPORTER` Kconfig option enables the key importer functionality.
```

> However, a provisioned device can update its firmware image from the Tinycrypt-based toolbox to firmware image that uses the PSA API based toolbox without requiring to be unprovisioned by using key importer functionality.

Reviewer 2: ```suggestion
A provisioned device can update its firmware image from the Tinycrypt-based toolbox to firmware image that uses the PSA API based toolbox without unprovisioning if the key importer functionality is used.
```

The key importer is an application initialization functionality that is called with kernel initialization priority before starting main.

> This functionality reads out the persistently stored Bluetooth Mesh data and if it finds keys stored by the Tinycrypt-based security toolbox, it imports them over the PSA API into the crypto library and stores the key identifiers in PSA API toolbox based format.

Reviewer 2: ```suggestion
This functionality reads out the persistently stored Bluetooth Mesh data and if it finds keys stored by the Tinycrypt-based security toolbox, it imports them over the PSA API into the crypto library and stores the key identifiers in a format based on the PSA API toolbox.
```

> Meanwhile, there is ability for a provisioned device to update its image with Tinycrypt based toolbox on image with the PSA API based toolbox without being unprovisioned.
> The following option :kconfig:option:`CONFIG_BT_MESH_KEY_IMPORTER` enables the key importer functionality.
> The key importer is an application initialization functionality that is called with kernel initialization priority before starting main.
> The functionality reads out the Bluetooth Mesh persistently stored data and if it finds keys stored by Tinycrypt based security toolbox it
> imports them over PSA API into crypto library and stores gotten key identifiers in PSA API toolbox based form.
> At the moment when application starts Bluetooth Mesh initialization, the persistent area already has stored data in the correct form.

Reviewer 2: ```suggestion
A provisioned device can update its image using the Tinycrypt-based toolbox and the PSA API based toolbox without being unprovisioned.
The :kconfig:option:`CONFIG_BT_MESH_KEY_IMPORTER` Kconfig option enables the key importer functionality.
The key importer is an application initialization functionality that is called with kernel initialization priority before starting main.
This functionality reads out the persistently stored Bluetooth Mesh data and if it finds keys stored by the Tinycrypt-based security toolbox, it imports them over the PSA API into the crypto library and stores the key identifiers in PSA API toolbox based format.
When the application starts Bluetooth Mesh initialization, the persistent area already has the stored data in the correct format.
```

Author: Rephrased a bit the first sentence to not contradict the previous subclause.

> When the application starts Bluetooth Mesh initialization, the persistent area already has the stored data in the correct format.

Reviewer: ```suggestion
Once the new firmware image starts Bluetooth Mesh initialization, the persistent area already has the stored data in the correct format due to use of key importer functionality.
```

> However, the key importer usage might add the potential vulnerability to the device. If device works with enabled key importer functionality
> and attacker gets ability to write arbitrary data in persistent memory then fake keys might be stored and will be imported to PSA crypto library after next device reset.

Reviewer 2: ```suggestion
Using the key importer might make the device vulnerable for attacks.
If the device works with the key importer functionality enabled and the attacker writes arbitrary data in the persistent memory, fake keys might be stored and imported to PSA crypto library after the next device reset.
```

> Complete the following steps to use the key importer functionality safely:

Reviewer: We should mention: Prevent flash readout by enabling flash read out protection (link)

> * The following steps should be done to use the key importer functionality safely:

Reviewer 2: ```suggestion
Complete the following steps to use the key importer functionality safely:
```

> Once the new firmware image starts Bluetooth Mesh initialization, the persistent area already has the stored data in the correct format due to use of key importer functionality.

Reviewer 2: ```suggestion
Once the new firmware image starts Bluetooth Mesh initialization, the persistent area already has the stored data in the correct format.
```

1. Update the images with the Tinycrypt-based toolbox and the PSA API based toolbox with the key importer feature enabled.
#. Reset the device to perform the key import after the devices in the network have successfully updated their images.
#. Update the images with the PSA API based security toolbox with the key importer feature disabled.

> 

Reviewer: I'm confused here. 
- Why one should "reset" the device? Does reset mean a factory reset or a reboot? 
- Opening line says: "Complete following ... to use key importer ...", and second bullet says disable key importer?

Author: Yes, it should be reboot. (I'll fix it)
Finally, key importer should be disabled since it adds vulnerability. It should be used only once to import keys and then it is not required at all.

>   * Update images with Tinycrypt based toolbox on images with the PSA API based toolbox with enabled the key importer feature.
>   * Reset device to perform key import after devices in the network have successfully updated their images.
>   * Update images with the PSA API based security toolbox but with disabled the key importer feature.

Reviewer 2: ```suggestion
1. Update the images with the Tinycrypt-based toolbox and the PSA API based toolbox with the key importer feature enabled.
#. Reset the device to perform the key import after the devices in the network have successfully updated their images.
#. Update the images with the PSA API based security toolbox with the key importer feature disabled.
```

> However, even these steps have been completed, the Tinycrypt based open key values is still possible to extract from the settings subsystem backend if there is access to it.
> The keys might still be compromised. It is strictly recommended to start the key refresh procedure according to specification for all existing keys.

Reviewer 2: ```suggestion
Even if you have completed these steps, the Tinycrypt-based open key values can be extracted from the settings subsystem backend if there is access to it.
The keys might still be compromised.
Start the key refresh procedure according to the specification for all existing keys.
```

  * Ambient light level gain

Adding up all entries, it is worth setting the cache size to minimum 71.

Security toolbox
----------------

Zephyr's Mesh security toolbox implementation uses third-party crypto libraries APIs (such as CMAC, AES-CCM, and HMAC-SHA-256) for implementing encryption and authentication functionality.

* The following options are available:

  * :kconfig:option:`CONFIG_BT_MESH_USES_MBEDTLS_PSA` - Enables use of the `Mbed TLS`_ PSA API based security toolbox (default option).
  * :kconfig:option:`CONFIG_BT_MESH_USES_TFM_PSA` - Enables use of the `Trusted Firmware M`_ PSA API based security toolbox (default option for platforms that support TF-M).
  * :kconfig:option:`CONFIG_BT_MESH_USES_TINYCRYPT` - Enables use of Tinycrypt-based security toolbox.
    Zephyr's Mesh operates with open key values, including storing them in the persistent memory.
    The Tinycrypt-based solution has worse security materials protection compared to others due to keeping keys in memory in open form.

> 	Tinycrypt is not recommended for future designs.

Reviewer 2: ```suggestion
	 Tinycrypt is not recommended for future designs.
```
This fixes the doc build error.

The Bluetooth Mesh security toolbox based on the `PSA Certified Crypto API`_ does not operate with open key values.
After Bluetooth Mesh receives an open key value, it immediately imports the key into the crypto library and receives the unique key identifier.
The key identifiers are used in the security toolbox and stored in the persistent memory.

> The crypto library is responsible for storing of the open key values in the Internal Trusted Storage (`PSA Certified Secure Storage API 1.0`_).

Reviewer: Just to avoid us trying to guess what crypto library does based on various Kconfig options enabled by end-user.

```suggestion
The crypto library is responsible for storing of the key values in the Internal Trusted Storage (`PSA Certified Secure Storage API 1.0`_).
```

The provisioned device cannot restore data from the persistent memory after firmware update.
If the image is changed over Mesh DFU, it is recommended to use :c:enumerator:`BT_MESH_DFU_EFFECT_UNPROV`.

A provisioned device can update its firmware image from the Tinycrypt-based toolbox to firmware image that uses the PSA API based toolbox without unprovisioning if the key importer functionality is used.
The :kconfig:option:`CONFIG_BT_MESH_KEY_IMPORTER` Kconfig option enables the key importer functionality.
The key importer is an application initialization functionality that is called with kernel initialization priority before starting main.
This functionality reads out the persistently stored Bluetooth Mesh data and if it finds keys stored by the Tinycrypt-based security toolbox, it imports them over the PSA API into the crypto library and stores the key identifiers in a format based on the PSA API toolbox.
Once the new firmware image starts Bluetooth Mesh initialization, the persistent area already has the stored data in the correct format.

> Using the key importer might make the device vulnerable for attacks.

Reviewer: Since device could be vulnerable already, key importer does not add new risk as such. Thus suggesting to rephrase:

```suggestion
Even after using the key importer the device can still be vulnerable to attacks during a period when device has not yet rebooted with firmware that uses key importer functionality. Two types of security risks are possible:
```

Author: This is not fully correct. Rephrased a little bit.

> If the device works with the key importer functionality enabled and the attacker writes arbitrary data in the persistent memory, fake keys might be stored and imported to PSA crypto library after the next device reset.
> 

Reviewer: ```suggestion
1. If the device is provided with a new image with the key importer functionality enabled , new image is not yet activated and the attacker writes arbitrary data in the persistent memory during this time by whichever methods, then fake keys might be imported to PSA crypto library after the next device reset (which activates the new firmware with key importer). This will lead to device getting provisioned to attackers network and attacker could read out the mesh state data from the device.

Mitigation: 
* Ensure that the device is protected from unauthorized writes to non-volatile storage.

2. Even after key importer imports the keys to internal trusted storage, the plain text values after left out in the flash until next garbage collection is triggered by storage backend. 

Mitigation: 
* Ensure that the device is protected from unauthorized reads (such as reading flash from programmer, or by using mcumgr shell commands) from the non-volatile storage.
* Execute key refresh procedure for all existing keys used on the entire network as soon as possible by excluding the compromised device, if any. Mechanism to determine if the device is compromised or not is up to the OEM developers.
```

Complete the following steps to use the key importer functionality safely:

1. Update the images with the Tinycrypt-based toolbox and the PSA API based toolbox with the key importer feature enabled.
#. Reboot the device to perform the key import after the devices in the network have successfully updated their images.
#. Update the images with the PSA API based security toolbox with the key importer feature disabled.

Even if you have completed these steps, the Tinycrypt-based open key values can be extracted from the settings subsystem backend if there is access to it.
The keys might still be compromised.

> Start the key refresh procedure according to the specification for all existing keys.

Reviewer: This is probably not needed now, if you accept the suggestion above.
